attentionbench/attention_utils.py

```diff
@@ -0,0 +1,124 @@
+
+    def get_name(self) -> str:
+        return f"attention_{self.B}x{self.M}x{self.N}x{self.K1}x{self.K2}x{self.dtype}"
+
+    def get_query_shape(self) -> str:
+        return f"{self.B}x{self.M}x{self.K1}x{self.dtype}"
+
+    def get_key_shape(self) -> str:
+        return f"{self.B}x{self.K2}x{self.K1}x{self.dtype}"
+
+    def get_value_shape(self) -> str:
+        return f"{self.B}x{self.K2}x{self.N}x{self.dtype}"
+
+    def get_output_shape(self) -> str:
+        return f"{self.B}x{self.M}x{self.N}x{self.dtype}"
+
+    def get_byte_count(self) -> int:
+        if "bf" in self.dtype:
+            bytes_per_element = int(self.dtype[2:]) // 8
+        else:
+            bytes_per_element = int(self.dtype[1:]) // 8
+        element_count = (
+            (self.B * self.M * self.K1)
+            + (self.B * self.K2 * self.K1)
+            + (self.B * self.K2 * self.N)
+        )
+        byte_count = element_count * bytes_per_element
+        return byte_count
+
+    def get_flops(self) -> int:
+        # We measure flops of the two matmuls only
+        qk_matmul_flops = 2 * self.M * self.K2 * self.K1
+        pv_matmul_flops = 2 * self.M * self.N * self.K2
+        total_flops = qk_matmul_flops + pv_matmul_flops
+        return total_flops
+
+
+def generate_mlir(config: AttentionConfig):
+    shapes = f"""\
+!dtype = {config.dtype}
+!Q     = tensor<{config.get_query_shape()}>
+!K     = tensor<{config.get_key_shape()}>
+!V     = tensor<{config.get_value_shape()}>
+!O     = tensor<{config.get_output_shape()}>
+"""
+
+    attn_kernel = """
+#Q = affine_map<(b, m, n, k1, k2) -> (b, m, k1)>
+#K = affine_map<(b, m, n, k1, k2) -> (b, k2, k1)>
+#V = affine_map<(b, m, n, k1, k2) -> (b, k2, n)>
+#O = affine_map<(b, m, n, k1, k2) -> (b, m, n)>
+
+func.func @main(%Q : !Q, %K : !K, %V : !V) -> !O {
+  %scale = arith.constant 1.0 : !dtype
+  %empty = tensor.empty() : !O
+  %O = iree_linalg_ext.attention { indexing_maps = [#Q, #K, #V, #O] }
+       ins(%Q, %K, %V, %scale : !Q, !K, !V, !dtype)
+       outs(%empty : !O) -> !O
+  return %O : !O
+}
+"""
+    mlir_template = shapes + "\n" + attn_kernel
+    return mlir_template
+
+
+def get_attention_flags() -> list[str]:
+    return []
+
+
+def compile_attention_config(
+    config: AttentionConfig, kernel_dir: Path, vmfb_dir: Path
+) -> tuple[AttentionConfig, Path, Optional[Path]]:
+    mlir_file = kernel_dir / (config.get_name() + ".mlir")
+    vmfb_file = vmfb_dir / (config.get_name() + ".vmfb")
+
+    # Generate mlir content
+    mlir_content = generate_mlir(config)
+
+    # Write MLIR content to file
+    with open(mlir_file, "w") as f:
+        f.write(mlir_content)
+
+    # TODO: Do not hardcode device information, instead pass it as a class
+    # Compile MLIR to vmfb
+    exec_args = [
+        "iree-compile",
+        # Input file
+        f"{mlir_file}",
+        # Output file
+        "-o",
+        f"{vmfb_file}",
+        # Backend: rocm
+        "--iree-hal-target-backends=llvm-cpu",
```

Comment:
Let's hardcode for mi300 for now (so rocm backend and gfx942 target). All our roofline plots (compute bound, memory bound) are based off this hardware

Reply:
Yeah, I was just using this to test on my local machine without a mi300